source/common/http/utility.cc

```diff
@@ -1494,5 +1499,57 @@ bool Utility::isValidRefererValue(absl::string_view value) {
   return true;
 }
 
+void handlePath(std::ostream& os, const HeaderString& value) {
+  std::string stripped = Utility::stripQueryString(value);
+  if (stripped != value.getStringView()) {
+    os << stripped << "?[redacted]";
+  } else {
+    os << value.getStringView();
+  }
+}
+
+void handleCookie(std::ostream& os, const HeaderString& value) {
+  bool first = true;
+  forEachCookie(value.getStringView(),
+                [&os, &first](absl::string_view k, absl::string_view) -> bool {
+                  if (!first) {
+                    os << ";";
+                  }
+                  first = false;
+                  os << k << "="
+                     << "[redacted]";
+
+                  // continue iterating until all cookies are processed.
+                  return true;
+                });
+}
+
+void Utility::dumpHeaderMap(std::ostream& os, const HeaderMap& headers, int indent_level) {
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.debug_include_sensitive_data")) {
+    headers.iterate([&os, spaces = spacesForLevel(indent_level)](
+                        const HeaderEntry& header) -> HeaderMap::Iterate {
+      os << spaces << "'" << header.key().getStringView() << "', '"
+         << header.value().getStringView() << "'\n";
+      return HeaderMap::Iterate::Continue;
+    });
+  } else {
+    headers.iterate([&os, spaces = spacesForLevel(indent_level)](
+                        const HeaderEntry& header) -> HeaderMap::Iterate {
+      os << spaces << "'" << header.key().getStringView() << "', '";
+      if (header.key() == Http::Headers::get().Path) {
```

Comment:
> This is where the spec would come in handy. 

What type of spec could it be?   

For masking three headers addressed in this PR currently, there was a need to write three different parsers: redacting query string, redacting cookie values, redacting complete header.  

Would the spec then provide means for user to 
1. apply pre-defined / fixed parsers to mask given headers?
2. add their own "custom masking" extensions written in C++ (like [access log formatters](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/formatter/formatter))?

To me (1) sounds reasonable, although only "redact complete header" is generic enough to be applied to any random header. 

I guess otherwise we end up in (2) allowing user to write application specific logic, to mask application specific headers. It sounds bit complicated and it might be unlikely people would utilize it that much. After all, masking debug log is quite a rare use case. 

 
> This change looks to be specific to your requirements. 

I thought the proposed headers were the obvious ones, which would apply to most, not only my requirements. 

But I agree, any `X-MyExampleApp-Token` type of header could contain sensitive information as well.

Reply:
Just a GitHub Issue with a short description of how a feature is going to be implemented is enough.

The key is flexibility. Envoy achieves this through extensions. I think the first step is create an extension point with API - interface with a method that takes header map as a parameter and returns formatted string. This can then plumbed into the header map.

Then you can build on top of this to add functionality that you need. I.e. list of headers that need to be redacted, boolean flag to redact query, cookies, etc. You can add regex matchers, whatever suits your purpose.

This is one of the possible solutions, other maintains may suggest simpler things. The Issue with the spec would help define the solution that meets community needs.